File: 김현경/[LV2] 피로도.py
def backtrack(k, visited, dungeons, cnt):
    max_cnt = cnt
    
    for i in range(len(dungeons)):
        if not visited[i] and k >= dungeons[i][0]:
            visited[i] = True
            max_cnt = max(max_cnt, backtrack(k - dungeons[i][1], visited, dungeons, cnt + 1))
            visited[i] = False          
    return max_cnt

def solution(k, dungeons):
    visited = [False] * len(dungeons)
    return backtrack(k, visited, dungeons, 0)

# from itertools import permutations
# def solution(k, dungeons):
#     max_cnt = 0
#     dungeon_permutations = list(permutations(dungeons))
    
#     for order in dungeon_permutations:
#         fatigue = k
#         cnt = 0
        
#         for min_fatigue, use_fatigue in order:
#             if fatigue >= min_fatigue:
#                 fatigue -= use_fatigue
#                 cnt += 1
#             else:
#                 break
#         max_cnt = max(max_cnt, cnt)
#     return max_cnt

File: 조성원/[LV2] 피로도.js
function exploreDungeons(k, dungeons, visited) {
  let maxDungeons = 0;

  for (let i = 0; i < dungeons.length; i++) {
    const [required, fatigue] = dungeons[i];
    if (visited[i] || k < required) continue;
    visited[i] = true;
    maxDungeons = Math.max(
      maxDungeons,
      1 + exploreDungeons(k - fatigue, dungeons, visited)
    );
    visited[i] = false;
  }

  return maxDungeons;
}

function solution(k, dungeons) {
  const visited = new Array(dungeons.length).fill(false);
  return exploreDungeons(k, dungeons, visited);
}
